test/support/sycl_alloc_utils.h

```diff
@@ -0,0 +1,145 @@
+// -*- C++ -*-
+//===----------------------------------------------------------------------===//
+//
+// Copyright (C) Intel Corporation
+//
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+// This file incorporates work covered by the following copyright and permission
+// notice:
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef __TEST_SYCL_ALLOC_UTILS_H
+#define __TEST_SYCL_ALLOC_UTILS_H
+
+#if TEST_DPCPP_BACKEND_PRESENT
+
+#include <list>
+#include <memory>
+
+#include "oneapi/dpl/pstl/hetero/dpcpp/sycl_defs.h"
+
+namespace TestUtils
+{
+////////////////////////////////////////////////////////////////////////////////
+//
+// RAII service class to allocate shared/device memory (USM)
+// Usage model"
+// 1. allocate USM memory and copying data to USM:
+//    usm_data_transfer<alloc_type, _ValueType> dt_helper(queue, first, count); 
+// or 
+//    usm_data_transfer<alloc_type, _ValueType> dt_helper(queue, std::begin(data), std::end(data));
+// or just allocate USM memory"
+//    usm_data_transfer<alloc_type, _ValueType> dt_helper(queue, count); 
+// 2. get a USM pointer by usm_data_transfer::get_data() and passed one into a parallel algorithm with dpc++ policy.
+// 3. Retrieve data back (in case of device allocation type) to the host for further checking result.
+//    dt_helper.retrieve_data(dest_host);
+//
+// Count of elements may be zero. In this case no USM memory will be allocated.
+// Also retrieve_data function will do nothing in this case.
+// This behavior is implemented based on tests use cases.
+//
+template<sycl::usm::alloc _alloc_type, typename _ValueType>
+class usm_data_transfer

```

Comment:
I still tend to think that the approach with class is overkill. I am not personally opposed to create our own RAII like objects but only if we have strong reasons to do that. Here I am not convinced we really have to do that.

And the argument about less function parameters on the caller side doesn't work for me because with having some "implicit" (from the user perspective) parameters make the call non-self contained and readability suffers.

I am not sure what to do with that now. Probably we could keeping for now and then revisit the design but again my opinion is we are not doing the right thing here.

Reply:
1) Other words, there is no errors and other blocking issues here and it can be merged, OK.
2) IMHO, I don't see any problems with a code pattern here. RAII pattern is absolutely applicable here.
3) Furthermore, I believe that we don't need waste time for the discussion here, because the test refactoring is coming soon and it will be discussed in that context in any case.